### PR TITLE
Hyacinth 833 lang dynamic field type

### PR DESCRIPTION
**Pull request by barmintor**

### CODE REVIEW

File: app/javascript/src/hyacinth_v1/components/digital_objects/common/DisplayField.js [hunk 1]
```diff
@@ -4,6 +4,8 @@ import InputGroup from '../../shared/forms/InputGroup';
 import Label from '../../shared/forms/Label';
 import PlainText from '../../shared/forms/inputs/PlainText';
 
+const randomId = (prefix) => `${prefix}${Math.random()}`.replace('.','');
```
Comment (elohanlon) [inline app/javascript/src/hyacinth_v1/components/digital_objects/common/DisplayField.js:7]: Do you mind changing this to the counter method instead?

Comment (barmintor) [reply app/javascript/src/hyacinth_v1/components/digital_objects/common/DisplayField.js:7]: Nope, sounds great.

File: spec/features/ui/v1/digital_objects/metadata/edit_spec.rb [hunk 2]
```diff
@@ -34,6 +34,35 @@
...
+        before do
+          page.find('.card-header', exact_text: 'Alternative Title')
+          card_header = page.find('.card-header', exact_text: 'Alternative Title')
+          within(card_header.sibling('.card-body')) do
+            # Selenium appears to be appending to current value with fill_in, so setting to blank first
+            page.find_field('Value').set updated_field_value
+            page.find_field('Value Language').set updated_field_lang
+          end
+          find_button('Update').click
+          # do a find to make sure scripts run
```
Comment (elohanlon) [inline spec/features/ui/v1/digital_objects/metadata/edit_spec.rb:53]: I don't remember what this comment means about scripts running.  Might be confusing for others too.

File: spec/features/ui/v1/digital_objects/metadata/edit_spec.rb [hunk 3]
```diff
@@ -34,6 +34,35 @@
...
+          find('.digital-object-interface')
+        end
+        it '' do
+          # do a find to make sure scripts run
```
Comment (elohanlon) [inline spec/features/ui/v1/digital_objects/metadata/edit_spec.rb:57]: Same comment here about what it means for scripts to run.

File: spec/features/ui/v1/digital_objects/metadata/edit_spec.rb (hunk 2 above)
Comment (barmintor) [reply spec/features/ui/v1/digital_objects/metadata/edit_spec.rb:53]: The find behavior makes Capybara wait for content to be rendered by an asynch script

File: spec/features/ui/v1/digital_objects/metadata/edit_spec.rb [hunk 4]
```diff
@@ -34,6 +34,35 @@
...
+          within(card_header.sibling('.card-body')) do
+            # Selenium appears to be appending to current value with fill_in, so setting to blank first
+            page.find_field('Value').set updated_field_value
+            page.find_field('Value Language').set updated_field_lang
+          end
+          find_button('Update').click
+          # do a find to make sure scripts run
+          find('.digital-object-interface')
+        end
+        it '' do
```
[the remark this answers is not in the document]
Comment (barmintor) [reply spec/features/ui/v1/digital_objects/metadata/edit_spec.rb:56]: Will do!

File: spec/models/concerns/digital_object_concerns/export_fields_behavior_spec.rb [hunk 5]
```diff
@@ -58,5 +58,28 @@
...
+      let(:value) { 'Ane Pleasant Satyre of the Thrie Estaitis' }
+      let(:field_export_profile) { FactoryBot.create(:field_export_profile, :for_alternative_title_field) }
+      let(:xmlns) { { 'mods' => 'http://example.org/' } }
+      let(:item) { FactoryBot.create(:item, :with_ascii_dynamic_field_data, value: value, value_lang: value_lang) }
+      # reload item for rehydrated term/language data
+      let(:loaded_item) { DigitalObject::Item.find(item.id) }
+      let(:exported_data) { loaded_item.render_field_export(field_export_profile) }
+      let(:exported_xml) { Nokogiri::XML(exported_data) }
+      before { FactoryBot.create(:export_rule, :for_alternative_title_field, field_export_profile: field_export_profile, dynamic_field_group: DynamicFieldGroup.first) }
+      it {
```
Comment (elohanlon) [inline spec/models/concerns/digital_object_concerns/export_fields_behavior_spec.rb:77]:
```suggestion
      it 'produces the expected XML' {
```
Related: I expected rubocop to complain here about using a curly brace for a multi-line block.

Comment (elohanlon) [reply spec/models/concerns/digital_object_concerns/export_fields_behavior_spec.rb:77]: The above suggestion is optional, btw.  Just personal preference.

File: app/javascript/src/hyacinth_v1/components/digital_objects/common/DisplayField.js (hunk 1 above)
Comment (elohanlon) [reply app/javascript/src/hyacinth_v1/components/digital_objects/common/DisplayField.js]: Thanks!

File: spec/support/dynamic_fields_helper.rb [hunk 6]
```diff
@@ -66,25 +66,37 @@ def load_sample_item_rights_fields!
     load_and_return!(rights_fields)
   end
 
-  def load_abstract_fields!
-    abstract_fields = {
+  def simple_value_field_data(string_key, **opts)
```
[the remark this answers is not in the document]
Comment (barmintor) [reply spec/support/dynamic_fields_helper.rb:69]: it's so comfy! But I think lost in the rebase is that the abstract field should loaded as a text area.

[the remark this answers is not in the document]
Comment (barmintor) [reply spec/support/dynamic_fields_helper.rb:69]: (which can be done with the opts hash)

File: spec/features/ui/v1/digital_objects/metadata/edit_spec.rb (hunk 2 above)
Comment (elohanlon) [reply spec/features/ui/v1/digital_objects/metadata/edit_spec.rb:53]: You might be able to do something like this instead (not sure if I have the exact syntax right):
```
Selenium::WebDriver::Wait.new(timeout: 15).until { find_button('Update') }
```

Comment (elohanlon) [reply spec/features/ui/v1/digital_objects/metadata/edit_spec.rb:53]: Well, actually, I guess if `find` waits anyway then the wait wrapper isn't adding anything.

**Review by elohanlon — APPROVED** [review]
Feel free to take or leave any suggestions.  PR approved!  Nice work.  👍

**Comment by barmintor** [comment]
I'm taking them all! Tweaks incoming, will merge after CI.

**Pull request merged**